### PR TITLE
with-cookie-auth: don't rely on the Host header

### PR DESCRIPTION
**Pull request by dunglas**
Relying on the `Host` HTTP header [is a bad practice that can lead to severe security vulnerabilities](https://dzone.com/articles/what-is-a-host-header-attack).

Even the code contained in this example can cause some damages:

    curl -H 'Host: attacker.com'  http://localhost:3000/profile

It forces the server to make an HTTP request to `http://attacker.com/api/profile.js`. It may among other things:

* reveal the origin IP address of your server, that you may want to hide (for instance when using DDOS protection systems such as CloudFlare)
* give many details about your platform including the User Agent of the HTTP client you use, or even [the exact version of the OS your server uses](https://www.infoq.com/articles/tcp-syn-security-unauthorized-os/)
* allow an attacker to amplify a DDOS attack using your infrastructure

This PR fixes the issue by storing the API URL in the configuration instead of relying on untrusted headers.

### CODE REVIEW

File: examples/with-cookie-auth/www/next.config.js [hunk 1]
```diff
@@ -0,0 +1,5 @@
+module.exports = {
+  publicRuntimeConfig: {
```
Comment (lfades) [inline examples/with-cookie-auth/www/next.config.js]: Use `env:` instead, https://nextjs.org/docs#build-time-configuration

**Comment by lfades** [comment]
@j0lv3r4 Can you review this PR 🙏

File: examples/with-cookie-auth/www/next.config.js (hunk 1 above)
Comment (Timer) [inline examples/with-cookie-auth/www/next.config.js]:
```suggestion
  env: {
```

File: examples/with-cookie-auth/www/pages/login.js [hunk 2]
```diff
@@ -2,16 +2,14 @@ import { Component } from 'react'
 import fetch from 'isomorphic-unfetch'
 import Layout from '../components/layout'
 import { login } from '../utils/auth'
+import getConfig from 'next/config'
```
Comment (Timer) [inline examples/with-cookie-auth/www/pages/login.js]:
```suggestion
```

File: examples/with-cookie-auth/www/pages/login.js [hunk 3]
```diff
@@ -2,16 +2,14 @@ import { Component } from 'react'
...
+
+const { publicRuntimeConfig } = getConfig()
```
Comment (Timer) [inline examples/with-cookie-auth/www/pages/login.js]:
```suggestion
```

File: examples/with-cookie-auth/www/pages/login.js [hunk 4]
```diff
@@ -2,16 +2,14 @@ import { Component } from 'react'
...
+const { API_URL } = publicRuntimeConfig
```
Comment (Timer) [inline examples/with-cookie-auth/www/pages/login.js]:
```suggestion
const { API_URL } = process.env
```

File: examples/with-cookie-auth/www/pages/profile.js [hunk 5]
```diff
@@ -3,6 +3,10 @@ import fetch from 'isomorphic-unfetch'
 import nextCookie from 'next-cookies'
 import Layout from '../components/layout'
 import { withAuthSync } from '../utils/auth'
+import getConfig from 'next/config'
```
Comment (Timer) [inline examples/with-cookie-auth/www/pages/profile.js]:
```suggestion
```

File: examples/with-cookie-auth/www/pages/profile.js [hunk 6]
```diff
@@ -3,6 +3,10 @@ import fetch from 'isomorphic-unfetch'
...
+
+const { publicRuntimeConfig } = getConfig()
```
Comment (Timer) [inline examples/with-cookie-auth/www/pages/profile.js]:
```suggestion
```

File: examples/with-cookie-auth/www/pages/profile.js [hunk 7]
```diff
@@ -3,6 +3,10 @@ import fetch from 'isomorphic-unfetch'
...
+const { API_URL } = publicRuntimeConfig
```
Comment (Timer) [inline examples/with-cookie-auth/www/pages/profile.js]:
```suggestion
const { API_URL } = process.env
```

**Review by Timer — CHANGES_REQUESTED** [review]

**Comment by j0lv3r4** [comment]
@lfades sure! 

Hey, @dunglas. This looks awesome. Thank you.

I think once the changes @Timer and @lfades suggested are done, this will be ready to merge.

**Comment by lfades** [comment]
@dunglas Did you test it after the changes ? afaik destructuring in `process.env` will only work in the server, but in the client it's `undefined`, so unless I'm wrong change the following:
```js
const { API_URL } = process.env
```
with
```js
const API_URL = process.env.API_URL
```

**Comment by timneutkens** [comment]
@lfades is correct, destructuring doesn't work, it has to be written in full as the value itself is replaced.

**Comment by dunglas** [comment]
Hi, no I just merged the suggestions for now. I'll test it (and fix it) asap.

**Comment by dunglas** [comment]
Unfortunately, it doesn't work. `process.env.API_URL` is `undefined` when using using build time config (it works when using runtime config). Any idea why?

**Comment by lfades** [comment]
@dunglas No idea, I'm going to clone the repo and check why

**Comment by lfades** [comment]
@dunglas @j0lv3r4 Try the example again, it should be working now 👍

**Comment by dunglas** [comment]
@lfades I confirm that it works properly. Thank you!

**Review by lfades — APPROVED** [review]

**Pull request merged**